Problem (Fix #1703): incorrect tree hashing in mls modules

yihuang opened the pull request:
Solution:
- Implement ratchet tree evolution
- Populate DirectPath for init_commit
- support apply the path secrets when apply commit message

leejw51crypto commented on chain-tx-enclave-next/mls/src/ciphersuite.rs:
```diff
@@ -77,12 +78,7 @@ impl Codec for ApplicationContext {
-    D::BlockSize: ArrayLength<u8>,
-    D::OutputSize: ArrayLength<u8>,
-{
+pub trait HkdfExt {
```
some comment about HkdftExt , can be replaced with fullword ?

tomtau replied:
it has a comment "Additional methods for Kdf".
It expands the "Hkdf" trait with extra methods defined in: https://github.com/mlswg/mls-protocol/blob/master/draft-ietf-mls-protocol.md#key-schedule

tomtau replied:
maybe it can reference it?

yihuang replied:
done

tomtau commented on chain-tx-enclave-next/mls/src/ciphersuite.rs:
```diff
@@ -296,15 +292,88 @@ impl CipherSuite {
+        }
+    }
+
+    pub fn secret_size(self) -> u16 {
```
key_pair secret?

yihuang replied:
The [go-mls](https://github.com/cisco/go-mls/blob/master/crypto.go#L69) used the name `SecretSize`, so

tomtau replied:
so what's this `secret_size` if not key pair secret size? the output hash length?

yihuang replied:
It’s used by expand label “path”, when generating path secrets.

tomtau replied:
https://github.com/mlswg/mls-protocol/blob/master/draft-ietf-mls-protocol.md#ratchet-tree-evolution

`KEM.Nsk` -- Nsk is the length of the secret key, so should be the same `keypair_secret_size`?

tomtau replied:
some of this changed, it may be good to put Git commit sha of spec this refers to

yihuang replied:
"Each ciphersuite has a `Derive-Key-Pair` function that maps octet strings of
length `Nsk` (a KEM-specific constant defined by HPKE) to HPKE key pairs", it seems they are still different?
And in go-mls, the [P521_AES256GCM_SHA512_P521](https://github.com/cisco/go-mls/blob/master/crypto.go#L108) has SecretSize 64, but the [DHKEM_P256](https://github.com/cisco/go-mls/blob/master/crypto.go#L292) has keyPairSecretSize 32.

yihuang replied:
> some of this changed, it may be good to put Git commit sha of spec this refers to

update the sha in lib.rs to refer to current master?

tomtau replied:
> "Each ciphersuite has a `Derive-Key-Pair` function that maps octet strings of
> length `Nsk` (a KEM-specific constant defined by HPKE) to HPKE key pairs", it seems they are still different?
> And in go-mls, the [P521_AES256GCM_SHA512_P521](https://github.com/cisco/go-mls/blob/master/crypto.go#L108) has SecretSize 64, but the [HPKEKEM](https://github.com/cisco/go-mls/blob/master/crypto.go#L292) has keyPairSecretSize 32.

66, not 32: https://github.com/cfrg/draft-irtf-cfrg-hpke/blob/master/draft-irtf-cfrg-hpke.md#key-encapsulation-mechanisms-kems-kem-ids

`shared_secret_size`? I'd need to read through more in detail whether that context is correct

tomtau replied:
as HPKE library doesn't implement P-521 or X448 yet, I'm not sure if it even distinguishes these sizes :) but still good to keep it consistent and distinguish them if that's in the spec

yihuang replied:
oh, I guess they have a bug here to [use DHKEM_P256 in ciphersuite P521_AES256GCM_SHA512_P521](https://github.com/cisco/go-mls/blob/master/crypto.go#L109)

leejw51crypto commented on chain-tx-enclave-next/mls/src/ciphersuite.rs:
```diff
@@ -153,13 +149,13 @@ where
 
 impl CipherSuite {
-    pub fn aead_key_len(&self) -> usize {
+    pub fn aead_key_len(self) -> usize {
```
comment of aead

tomtau commented on chain-tx-enclave-next/mls/src/ciphersuite.rs:
```diff
@@ -296,15 +292,88 @@ impl CipherSuite {
+        }
+    }
+
+    pub fn derive_private_key(
```
from Ciphersuites def?

tomtau replied:
probably fixme or change to use https://github.com/cfrg/draft-irtf-cfrg-hpke/blob/master/draft-irtf-cfrg-hpke.md#derivekeypair-derive-key-pair

tomtau replied:
the problem here is that not all secrets from that "key pair expand label" may be valid, so the unmarshal may panic

yihuang replied:
added a FIXME for now, the go version seems don't handle this currently.

tomtau replied:
that definitions isn't longer in the spec?

tomtau replied:
that expand_label with "key pair"?

yihuang replied:
"This function is defined to be HPKE's `DeriveKeyPair` function.", I think it refers to the [hpke spec](https://github.com/cfrg/draft-irtf-cfrg-hpke/blob/master/draft-irtf-cfrg-hpke.md#derivekeypair-derive-key-pair) now.

yihuang replied:
The DeriveKeyPair function should ideally implemented in hpke library?

yihuang replied:
the hpke library just implemented derive_keypair, we changed to use that now.

leejw51crypto commented on chain-tx-enclave-next/mls/src/ciphersuite.rs:
```diff
@@ -278,7 +274,7 @@ impl CipherSuite {
     /// TODO: use generic array?
     pub fn encrypt_group_info(
-        &self,
+        self,
```
just &[u8] is fastest, not necessary to use generic array

tomtau replied:
generic array is meant rather for welcome_key, welcome_nonce ... not the (variable-length) plaintext input

leejw51crypto commented on chain-tx-enclave-next/mls/src/ciphersuite.rs:
```diff
@@ -296,15 +292,88 @@ impl CipherSuite {
 
     /// TODO: use generic array?
-    pub fn hash(&self, data: &[u8]) -> Vec<u8> {
+    pub fn hash(self, data: &[u8]) -> Vec<u8> {
```
this is fine, not necessary to use generic array

tomtau replied:
generic array would prevent some problems at compile-time, because there are various `Vec<u8>` (hashes, secrets, nonces, pubkeys...) which can be quite error-prone

tomtau commented on chain-tx-enclave-next/mls/src/group.rs:
```diff
@@ -174,19 +175,38 @@ impl GroupAux {
         let positions = updated_tree.update(&add_proposals, &[], &[]);
-        let commit_secret = vec![0; self.tree.cs.hash_len()];
+
+        // update path secrets
+        let mut leaf_secret = vec![0; self.tree.cs.hash_len()];
+        rand::thread_rng().fill_bytes(&mut leaf_secret);
+        let (path_nodes, parent_hash, commit_secret) =
+            updated_tree.encap(&updated_group_context.get_encoding(), leaf_secret);
+
+        // update keypackage's parent_hash extension
+        self.owned_kp
+            .keypackage
+            .payload
+            .put_extension(&ext::ParentHashExt(parent_hash));
+        self.owned_kp.update_signature();
+        updated_tree.set_my_package(self.owned_kp.keypackage.clone());
```
> The generator of the Commit starts by using the HPKE secret key "leaf_hpke_secret" associated with the new leaf KeyPackage (see {{key-packages}}) to compute "path_secret[0]"

tomtau replied:
init key is to be updated -- https://github.com/cisco/go-mls/blob/master/state.go#L370 ?

tomtau replied:
> When used as InitKeys, KeyPackages are intended to be used only once and SHOULD NOT be reused except in case of last resort.

yihuang replied:
done

leejw51crypto commented on chain-tx-enclave-next/mls/src/key.rs:
```diff
@@ -100,6 +104,14 @@ impl HPKEPrivateKey {
         &self.0
     }
+
+    pub fn unmarshal(secret: &[u8]) -> Result<Self, HpkeError> {
```
consider `decode` for unmarshal, because we're using `encode`, `decode` terms

tomtau replied:
these are keywords used in the spec

leejw51crypto commented on chain-tx-enclave-next/mls/src/keypackage.rs:
```diff
@@ -81,6 +81,19 @@ impl KeyPackagePayload {
     }
 
+    /// insert or update extension
+    pub fn put_extension<T: MLSExtension>(&mut self, ext: &T) {
```
pub fn put_extension(&mut self, ext: &MLSExtension) { 
is simpler

devashishdxt replied:
Maybe use `pub fn put_extension(&mut self: ext: &impl MLSExtension)`?

yihuang replied:
done

leejw51crypto commented on chain-tx-enclave-next/mls/src/tree.rs:
```diff
@@ -350,7 +390,148 @@ impl Tree {
+    /// update path secrets
+    ///
+    /// returns: direct path nodes, parent_hash, root_secret
+    pub fn encap(
```
full word for `encap`

tomtau replied:
or something different -- `Encap` has already a defined meaning in HPKE, but here it's something else

yihuang replied:
maybe "evolve"

leejw51crypto commented on chain-tx-enclave-next/mls/src/tree.rs:
```diff
@@ -350,7 +390,148 @@ impl Tree {
+            .get_encoding(),
+        ),
+        Node::Parent(pn) => {
+            let left_index = index.left().unwrap();
```
expect()?

yihuang replied:
done

tomtau commented on chain-tx-enclave-next/mls/src/key.rs:
```diff
@@ -88,6 +91,7 @@ impl Codec for HPKEPublicKey {
 
 /// p-256 private key
 /// used for obtaining the initial sealed secrets (HPKE)
+#[derive(Clone)]
 pub struct HPKEPrivateKey(<hpke::kex::DhP256 as hpke::KeyExchange>::PrivateKey);
```
is this needed to be clone?

yihuang replied:
removed

yihuang replied:
The `Clone` is added back, because we need to store private keys inside tree, and tree need to be able to be cloned.

tomtau commented on chain-tx-enclave-next/mls/src/tree.rs:
```diff
@@ -350,7 +390,148 @@ impl Tree {
+    pub fn encap(
+        &mut self,
+        ctx: &[u8],
+        leaf_secret: Vec<u8>,
```
SecretVec?

yihuang replied:
done

tomtau commented on chain-tx-enclave-next/mls/Cargo.toml:
```diff
@@ -19,6 +19,7 @@ aead = "0.2.0"
 rand = "0.7"
 ra-client = { path = "../enclave-ra/ra-client" }
 subtle = "2.2.3"
+rand_core = "0.5"
```
why rand_core? isn't that exposed from rand?

yihuang replied:
removed

yihuang commented:
support apply path secrets when apply commit message now, existing unit tests passed.

yihuang commented:
> maybe splitting the tree into private and public portions similarly to https://github.com/cisco/go-mls/pull/71/files#diff-53faa623cefecaf31c5a72febc6526c9R363 ?
> need to yet read through the code in more detail, as some spec parts changed a bit

how about do this in another pr, I think we need better code change to make the tree implementation more robust [sketched here](https://gist.github.com/yihuang/3155363033f6b7c4c17f4b9d42894df2)

tomtau commented:
> > maybe splitting the tree into private and public portions similarly to https://github.com/cisco/go-mls/pull/71/files#diff-53faa623cefecaf31c5a72febc6526c9R363 ?
> > need to yet read through the code in more detail, as some spec parts changed a bit
> 
> how about do this in another pr, I think we need more code changes to make the tree implementation more robust [sketched here](https://gist.github.com/yihuang/3155363033f6b7c4c17f4b9d42894df2)

should be ok, I'll try to do a final read of this PR later today -- these other code improvements can be in a different PR

tomtau commented on chain-tx-enclave-next/mls/src/tree.rs:
```diff
@@ -335,34 +441,264 @@ impl Tree {
+}
+
+/// no blank nodes return
+fn resolve(nodes: &[Node], index: NodeSize) -> Vec<NodeSize> {
```
could be good to link "Ratchet Tree Nodes" section -- I was trying to search "resolve" where it is defined

yihuang replied:
done

tomtau commented on chain-tx-enclave-next/mls/src/tree.rs:
```diff
@@ -335,34 +441,264 @@ impl Tree {
+    }
+
+    /// returns (leaf_parent_hash, commit_secret)
+    pub fn apply_path_secrets(
```
"Synchronizing Views of the Tree"?

yihuang replied:
done

tomtau commented on chain-tx-enclave-next/mls/src/tree.rs:
```diff
@@ -335,34 +441,264 @@ impl Tree {
+            self.nodes[node_index.0].merge_public(path_node.public_key.clone());
+        }
+
+        // Find overlap and decrypt the path secret
+        let (overlap, mut last_secret) =
+            self.decrypt_path_secrets(sender, ctx, path_nodes, leaf_private_key);
+
+        // implant the path secret
+        self.nodes[overlap.0]
+            .parent_node_mut()
+            .expect("implant invalid node")
+            .set_secret(self.cs, &last_secret);
+        for node_index in overlap.direct_path(leaf_len).iter() {
+            last_secret = self
+                .cs
+                .expand_label(&last_secret, vec![], "path", &[], self.cs.secret_size())
+                .expect("invalid length");
+            self.nodes[node_index.0]
+                .parent_node_mut()
+                .expect("implant invalid node")
+                .set_secret(self.cs, &last_secret);
+        }
```
A bit confused about this part -- my assumption is, as this is called by the recipient in processing commit in group init, that this is "Synchronizing Views of the Tree".
There's one step:
> The recipient SHOULD verify that the received public keys agree with the public keys derived from the new path_secret values.

I'm more sure whether that's done?

yihuang replied:
Changed the `set_secret` to `set_secret_verify_public`, the verification is done inside. the public key is updated in `merge_public_key`.

yihuang commented:
> looks ok -- there may be one step missing or I missed that when reading the code -- in general, it'd be good to add pointers to spec in the more involved parts

Fixed, also changed the error types to be more specific.

yihuang commented:
> one small clippy fix: https://travis-ci.org/github/crypto-com/chain/jobs/699644837#L1837

done

tomtau commented:
bors retry
